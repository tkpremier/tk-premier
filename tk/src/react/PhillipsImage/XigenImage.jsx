import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { includes } from 'lodash';
import { has } from 'lodash/fp';

const transformationDictionary = {
  AuctionCatalogBuy: '&width=200',
  AuctionHighlightsGalleryInPage: '&width=590&height=472&sizemode=fitpadded&align=center',
  AuctionHighlightsGalleryModal: '&width=900&height=500&sizemode=fitpadded&align=center',
  AuctionInlinePopup: '/464/535/true/true/true/fitpadded/right',
  AuctionLotsView: '/464/593/true/true/true/fitpadded/bottom',
  MyPhillipsLotsView: '/220/281/true/true/true/fitpadded/left',
  HomePageCarousel: '/220/281/true/true/true/fitpadded/bottom',
  AuctionLotsViewSmaller: '/172/220/true/true/true/fitpadded/bottom',
  AuctionMidBannerImg: '&width=1039&height=831',
  AuctionPastBannerGallery: '&width=308&height=222',
  AuctionPastGallery: '&width=308&height=222&sizemode=fitpadded&align=right',
  AuctionThumbOnLocation: '&width=250',
  LotDetailAltImageFullSize: '&width=410&height=550&padx=false&pady=true',
  LotDetailAlternateThumbs: '&width=120&height=400',
  LotDetailAlternateThumbsMainImage: '/120/400',
  LotDetailImageCustomSize: '/{0}/{1}/true/true/true/fitpadded/top',
  LotDetailMainImage: '/605/550/false/false/false',
  LotDetailMainImageSmaller: '/410/550/false/false/false',
  LotDetailSideImage: '/325/220/false/false/false',
  LotDetailZoomImage: '/2000/4000',
  LotDetailZoomImageAlt: '&width=2000&height=4000',
  PastAuctionBannerImage: '&width=960&height=767',
  SearchResultsLots: '&width=175&height=170&padx=true&pady=true',
  VideoImage: '&width=219&height=219&sizemode=fitpadded&align=center&background-color=#ccc;',
  // Curated Auction Transformations
  SingleCell: '/464/593/true/true/true/fitpadded/bottom',
  TwoColumnsOneRow: '/928',
  ThreeColumns: '/1392',
  ThreeColumnsOneRow: '/1392/593',
  TwoRowsTwoColumns: '/928/1186',
  TwoRows: '/464/1186',
  TwoLotsOneRow: '/650/593'
};

class XigenImage extends Component {
  constructor() {
    super();
    this.imgRef = null;
    this.handleLoad = this.handleLoad.bind(this);
    this.state = {
      loaded: false
    };
  }
  componentDidMount() {
    if (this.imgRef.complete && !this.state.loaded) {
      this.handleLoad(this.imgRef);
    }
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.imagePath !== this.props.imagePath) {
      this.setState({
        loaded: false
      });
    }
  }

  getImageSource() {
    const diskStationTransforms = [
      'LotDetailAltImageFullSize',
      'LotDetailAlternateThumbs',
      'AuctionHighlightsGalleryInPage',
      'AuctionHighlightsGalleryModal',
      'AuctionPastGallery'
    ];
    let imageUrl = this.props.imagePath;
    if (this.props.transformation) {
      const transformation = transformationDictionary[this.props.transformation];
      if (typeof transformation === 'undefined') {
        console.warn(`transformation ${this.props.transformation} does not exist in dictionary`);
      } else {
        imageUrl = `${imageUrl}${transformation}`;
        if (includes(diskStationTransforms, this.props.transformation)) {
          imageUrl = `https://www.phillips.com/Xigen/image.ashx?path=${imageUrl}`;
        }
      }
    } else {
      imageUrl = `${imageUrl}/${this.props.width}/${this.props.height}`;
    }

    return imageUrl;
  }

  handleLoad(e) {
    const event = has('target')(e) ? e : { 'target': e, 'type': 'load' };
    if (!this.state.loaded) {
      this.props.onLoad(event);
      this.setState({ loaded: true });
    }
  }

  render() {
    const src = this.getImageSource();
    return (
      <img
        alt={this.props.alt}
        className="xigen-image"
        data-pin-url={src}
        onError={this.props.onError}
        onLoad={this.handleLoad}
        src={src}
        title={this.props.alt}
        ref={(el) => { this.imgRef = el; }}
      />
    );
  }
}


XigenImage.defaultProps = {
  transformation: false,
  constrain: false,
  padX: false,
  padY: false,
  sizeMode: null,
  align: null,
  backgroundColor: null
};

XigenImage.propTypes = {
  imagePath: PropTypes.string.isRequired,
  alt: PropTypes.string,
  onLoad: PropTypes.func,
  onError: PropTypes.func
};

export default XigenImage;
