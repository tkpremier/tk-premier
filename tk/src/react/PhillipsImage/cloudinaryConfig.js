import { Cloudinary } from 'cloudinary-core';

const cloudinaryConfig = Cloudinary.new({
  cloud_name: 'phillips-assets',
  private_cdn: true,
  secure: true,
  secure_distribution: 'assets.phillips.com',
  cname: 'assets.phillips.com'
});

export default cloudinaryConfig;
