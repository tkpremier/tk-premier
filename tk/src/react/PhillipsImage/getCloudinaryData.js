/**
 *
 * @param {string} url url string to parse;
 *
 * @typedef imageData
 * @type {object}
 * @property {string} publicId - original public Id before transform
 * @property {number} cloudinaryVersion - number of current version
 * @property {string} transform - transformation name
 *
 * @return {imageData} - returns the imageData
 */
const domain = 'https://assets.phillips.com/image/upload/';

const getCloudinaryImageData = (url = '') => {
  try {
    const transformation = url.split(domain)[1].split('/')[0];
    const cloudinaryVersion = url.split(`${transformation}/`)[1].split('/')[0];
    const publicId = url.split(`${cloudinaryVersion}`)[1];
    return {
      cloudinaryVersion,
      publicId,
      transformation
    };
  } catch (e) {
    console.log('There was an error getting image data.  Make sure URL is cloudinary, phillips, and correct: ', e);
    return {
      cloudinaryVersion: '1',
      publicId: '/',
      transformation: 'Error'
    };
  }
};

export default getCloudinaryImageData;
