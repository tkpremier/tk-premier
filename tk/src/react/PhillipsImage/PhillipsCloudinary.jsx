import React, { Component } from 'react';
import { has } from 'lodash/fp';
import PropTypes from 'prop-types';
import cloudinaryConfig from './cloudinaryConfig';

const cloudinaryTransformations = {
  AuctionHighlightsGalleryModal: 'Website_AuctionHighlightsGalleryModal',
  AuctionInlinePopup: 'Website_AuctionPageLot',
  AuctionLotsView: 'Website_AuctionPageLot',
  AuctionThumb: 'Website_Auction_Thumb_dev4',
  EditorialHubFeaturedMain: 'editorial_image_featured-main',
  EditorialHubFullWidth: 'editorial_image_full-width',
  EditorialHub: 'editorial_image',
  EditorialHubVideoThumb: 'editorial_hub_feed_video_dev3',
  HomePageCarousel: 'Website_HomePageCarousel',
  LotDetailMainImage: 'Website_LotDetailMainImage',
  LotDetailZoomImage: 'Website_LotDetailZoomImage',
  LotDetailAlternateThumbs: 'Website_LotDetailThumb',
  LotDetailAlternateThumbsMainImage: 'Website_LotDetailThumb',
  LotDetailAltImageFullSize: 'Website_LotDetailMainImage',
  LotDetailZoomImageAlt: 'Website_LotDetailZoomImage',
  // Curated Auction Transformations
  SingleCell: 'Website_CuratedAuction_SingleCell_NoPad_dev2',
  // Curated Auction Containers
  ContainerTworowsTwocolumnsFeatureLeft: 'Website_CuratedAuction_TwoColumns_dev2',
  ContainerTworowsTwocolumnsFeatureRight: 'Website_CuratedAuction_TwoColumns_dev2',
  ContainerTworowsTwocolumnsFeatureCenter: 'Website_CuratedAuction_TwoColumns_dev2',
  ContainerTworowsTwocolumnsTopRight: 'Website_CuratedAuction_SingleCell_NoPad_dev2',
  ContainerTworowsTwocolumnsBtmRight: 'Website_CuratedAuction_SingleCell_NoPad_dev2',
  ContainerTworowsTwocolumnsTopLeft: 'Website_CuratedAuction_SingleCell_NoPad_dev2',
  ContainerTworowsTwocolumnsBtmLeft: 'Website_CuratedAuction_SingleCell_NoPad_dev2',
  ContainerTworowsThreecolumnsTopLeft: 'Website_CuratedAuction_SingleCell_NoPad',
  ContainerTworowsThreecolumnsBottomLeft: 'Website_CuratedAuction_SingleCell_NoPad',
  ContainerTworowsThreecolumnsTopMiddle: 'Website_CuratedAuction_SingleCell_NoPad',
  ContainerTworowsThreecolumnsBottomMiddle: 'Website_CuratedAuction_SingleCell',
  ContainerTworowsThreecolumnsTopRight: 'Website_CuratedAuction_SingleCell_NoPad',
  ContainerTworowsThreecolumnsBottomRight: 'Website_CuratedAuction_SingleCell_NoPad',
  ContainerTworowsThreecolumnsFeatureLeft: 'Website_CuratedAuction_TwoRows',
  ContainerTworowsThreecolumnsFeatureMiddle: 'Website_CuratedAuction_TwoRows',
  ContainerTworowsThreecolumnsFeatureRight: 'Website_CuratedAuction_TwoRows',
  OnerowTwocolumnsLeft: 'Website_CuratedAuction_TwoColumnsOneRowNoPad',
  OnerowTwocolumnsRight: 'Website_CuratedAuction_TwoColumnsOneRowNoPad',
  TwoColumnsOneRowNoPad: 'Website_CuratedAuction_TwoColumnsOneRowNoPad',
  TwoColumnsOneRow: 'Website_CuratedAuction_TwoColumnsOneRowNoPad',
  ThreeColumns: 'Website_CuratedAuction_ThreeColumns',
  ThreeColumnsOneRow: 'Website_CuratedAuction_ThreeColumns',
  TwoRowsTwoColumns: 'Website_CuratedAuction_TwoRowsTwoColumns',
  TwoRows: 'Website_CuratedAuction_TwoRows',
  TwoLotsOneRow: 'Website_CuratedAuction_TwoLotsOneRow'
};

export const getCloudinaryTransformation = (transformation) => {
  if (typeof cloudinaryTransformations[transformation] === 'undefined') {
    console.error(`Cloudinary Transformation ${transformation} does not exist`);
    return 'error';
  }
  return cloudinaryTransformations[transformation];
};

class PhillipsCloudinary extends Component {
  constructor(props) {
    super(props);
    this.imgRef = null;
    this.handleLoad = this.handleLoad.bind(this);
    this.state = {
      loaded: false
    };
  }

  componentDidMount() {
    if (this.imgRef.complete && !this.state.loaded) {
      this.handleLoad(this.imgRef);
    }
  }

  componentWillReceiveProps(nextProps) {
    if (this.props.publicId !== nextProps.publicId) {
      this.setState({
        'loaded': false
      });
    }
  }

  getImageSource() {
    // set protocol dynamically ?
    // switch imageSource to cname once configured
    return this.props.useTransformation
      ? cloudinaryConfig.url(
        this.props.publicId,
        {
          transformation: getCloudinaryTransformation(this.props.transformation),
          version: this.props.version
        }
      )
      : cloudinaryConfig.url(
        this.props.publicId,
        {
          width: this.props.width,
          height: this.props.height,
          crop: this.props.crop
        }
      );
  }

  handleLoad(e) {
    const event = has('target')(e) ? e : { 'target': e, 'type': 'load' };
    if (!this.state.loaded) {
      this.props.onLoad(event);
      this.setState({ loaded: true });
    }
  }

  render() {
    const src = this.getImageSource();
    if (!this.props.useTransformation) {
      console.log('src: ', src);
    }
    return (
      <img
        alt={this.props.alt}
        data-pin-url={src}
        onError={this.props.onError}
        onLoad={this.handleLoad}
        src={src}
        title={this.props.alt}
        ref={(el) => { this.imgRef = el; }}
      />
    );
  }
}

PhillipsCloudinary.defaultProps = {
  alt: 'no alt text provided',
  width: 250,
  height: 320,
  crop: 'scale',
  version: '1',
  angle: 0,
  effect: '',
  overlay: '',
  transformation: 'AuctionLotsView',
  useTransformation: true
};

PhillipsCloudinary.propTypes = {
  alt: PropTypes.string,
  publicId: PropTypes.string.isRequired,
  version: PropTypes.string,
  width: PropTypes.number,
  height: PropTypes.number,
  crop: PropTypes.string,
  transformation: PropTypes.string,
  useTransformation: PropTypes.bool,
  onLoad: PropTypes.func.isRequired,
  onError: PropTypes.func.isRequired
};

export default PhillipsCloudinary;
