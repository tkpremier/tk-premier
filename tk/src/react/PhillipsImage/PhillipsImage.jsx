import 'fetch-ponyfill';
import React, { useEffect, useRef, useState } from 'react';
import isNull from 'lodash/isNull';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import cloudinaryConfig from './cloudinaryConfig';
import { getCloudinaryTransformation } from './PhillipsCloudinary';
import XigenImage from './XigenImage';

const placeholder = 'https://phillips.vo.llnwd.net/v1/web_prod/images/placeholders/item_placeholder.png';
const Loading = () => {
  const ref = useRef(null);
  const [height, setHeight] = useState('133.33px');
  useEffect(() => {
    if (!isNull(ref.current)) {
      const newWidth = ref.current.clientWidth;
      const newHeight = newWidth * (2 / 3);
      setHeight(`${parseInt(newHeight, 10)}px`);
    }
  }, []);
  return (
    <div className="image-loader" ref={ref} style={{ height, zIndex: 10 }}><div className="signal" /></div>
  );
};
const Image = (props) => {
  const [loaded, setLoaded] = useState(false);
  const ref = useRef(null);
  const handleLoad = (event) => {
    props.onLoad(event);
    if (!loaded) {
      setLoaded(true);
    }
  };
  useEffect(() => {
    const domNode = ref.current;
    const effectHandleLoad = handleLoad;
    const onComplete = (event) => {
      if (!loaded) {
        props.onLoad(event);
        setLoaded(true);
      }
    };
    if (!isNull(domNode)) {
      domNode.addEventListener('complete', onComplete);
      if (ref.current.complete && !loaded) {
        effectHandleLoad({ target: ref.current, type: 'load' });
        setLoaded(true);
      }
    }
    return () => {
      domNode.removeEventListener('complete', onComplete);
    }
  }, []);
  return !isNull(props.imgSrc)
    ? (
      <img
        alt={props.alt}
        data-pin-url={props.imgSrc}
        loading={props.loading}
        onError={props.onError}
        onLoad={handleLoad}
        src={props.imgSrc}
        ref={ref}
      />
    )
    : (
      <XigenImage
        onLoad={handleLoad}
        onError={props.onError}
        {...props}
      />
    );
};

Image.defaultProps = {
  alt: '',
  loading: 'lazy',
  imgSrc: placeholder
}
Image.propTypes = {
  alt: PropTypes.string,
  imgSrc: PropTypes.string,
  loading: PropTypes.string,
  onError: PropTypes.func.isRequired,
  onLoad: PropTypes.func.isRequired
}

const PhillipsImage = (props) => {
  const [loading, setLoading] = useState(true);
  const [error, setError] = useState(false);
  const handleLoad = (event) => {
    if (props.onLoad) {
      props.onLoad(event);
    }
    if (loading) {
      setLoading(!loading);
    }

  };
  const handleError = () => {
    if (!error) {
      if (props.onError) {
        props.onError();
      }
      setLoading(false);
      setError(true);
    }
  }
  const imgSrc = props.cloudinary
    ? error
      ? placeholder
      : props.useTransformation
        ? cloudinaryConfig.url(
          props.imagePath,
          {
            transformation: getCloudinaryTransformation(props.transformation),
            version: props.version
          }
        )
        : cloudinaryConfig.url(
          props.imagePath,
          {
            width: props.width,
            height: props.height,
            crop: props.crop
          })
    : null;
  return (
    <div className={classNames('phillips-image', props.className)}>
      {loading
        ? <Loading />
        : null
      }
      <Image
        {...props}
        imgSrc={imgSrc}
        onError={handleError}
        onLoad={handleLoad}
      />
    </div>
  );
}

PhillipsImage.defaultProps = {
  alt: '',
  className: '',
  cloudinary: true,
  constrain: false,
  crop: 'scale',
  curatedView: false,
  description: 'no description',
  height: 280,
  hideLoader: false,
  loading: 'lazy',
  onError: null,
  onLoad: null,
  padX: false,
  padY: false,
  sizeMode: 'fitpadded',
  transformation: 'AuctionLotsView',
  useTransformation: true,
  version: '1',
  width: 220
};

PhillipsImage.propTypes = {
  alt: PropTypes.string,
  className: PropTypes.string,
  cloudinary: PropTypes.bool,
  crop: PropTypes.string,
  curatedView: PropTypes.bool,
  onLoad: PropTypes.func,
  onError: PropTypes.func,
  loading: PropTypes.string,
  hideLoader: PropTypes.bool,
  description: PropTypes.string,
  transformation: PropTypes.string,
  imagePath: PropTypes.string.isRequired,
  height: PropTypes.number,
  constrain: PropTypes.bool,
  padX: PropTypes.bool,
  padY: PropTypes.bool,
  sizeMode: PropTypes.string,
  version: PropTypes.string,
  useTransformation: PropTypes.bool,
  width: PropTypes.number
};

export default PhillipsImage;
