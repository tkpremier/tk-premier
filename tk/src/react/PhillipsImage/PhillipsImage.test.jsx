import React from 'react';
import renderer from 'react-test-renderer';
import PhillipsImage from './PhillipsImage';

describe('PhillipsImage', () => {
  it('src should be /690/550/false/false/false/fitpadded/top', () => {
    const phillipsImage = renderer.create(
      <PhillipsImage imagePath={'/xigen/lotimg/Gino-Sarfatti/NY050117/1'} width={690} height={550} align={'top'} />
    );
    const tree = phillipsImage.toJSON();
    expect(tree).toMatchSnapshot();
  });

});