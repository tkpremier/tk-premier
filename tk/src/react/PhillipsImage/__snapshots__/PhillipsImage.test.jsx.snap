// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PhillipsImage src shoud be /690/550/false/false/false/fitpadded/top 1`] = `
<img
  alt={undefined}
  src="/fakeurl"
/>
`;

exports[`PhillipsImage src should be /690/550/false/false/false/fitpadded/top 1`] = `
<div
  className="phillips-image"
>
  <div
    className="image-loader"
    style={
      Object {
        "height": "550px",
        "width": "690px",
      }
    }
  >
    <span
      className="signal"
    />
  </div>
  <img
    alt="no alt text provided"
    data-pin-url="https://assets.phillips.com/image/upload/t_Website_AuctionPageLot/v1/xigen/lotimg/Gino-Sarfatti/NY050117/1"
    onError={[Function]}
    onLoad={[Function]}
    src="https://assets.phillips.com/image/upload/t_Website_AuctionPageLot/v1/xigen/lotimg/Gino-Sarfatti/NY050117/1"
  />
</div>
`;
